WcaOnRails/spec/controllers/delegate_reports_controller_spec.rb

```diff
@@ -0,0 +1,87 @@
+require 'rails_helper'
+
+describe DelegateReportsController do
+  let(:comp) { FactoryGirl.create(:competition, :with_delegate) }
+
+  context "not logged in" do
+    it "redirects to sign in" do
+      get :show, competition_id: comp.id
+      expect(response).to redirect_to(new_user_session_path)
+    end
+  end
+
+  context "logged in as a regular user" do
+    sign_in { FactoryGirl.create(:user) }
+
+    it "redirects to home page" do
+      get :show, competition_id: comp.id
+      expect(response).to redirect_to(root_url)
+    end
+  end
+
+  context "logged in as a regular delegate" do
+    sign_in { FactoryGirl.create(:delegate) }
+
+    it "redirects to home page" do
+      get :edit, competition_id: comp.id
+      expect(response).to redirect_to(root_url)
+    end
+
+    it "redirects to home page" do
+      patch :update, competition_id: comp.id
+      expect(response).to redirect_to(root_url)
+    end
+  end
+
+  context "logged in as THE delegate" do
+    before :each do
+      sign_in comp.delegates.first
+    end
+
+    it "can view edit page" do
+      get :edit, competition_id: comp.id
+      expect(response.status).to eq 200
+    end
+
+    it "can edit report" do
+      post :update, competition_id: comp.id, delegate_report: { remarks: "My new remarks" }
+      expect(response).to redirect_to(delegate_report_edit_path(comp))
+      comp.reload
+      expect(comp.delegate_report.remarks).to eq "My new remarks"
+    end
+
+    it "can edit report before comp is over, but cannot post report" do
+      # Update comp to be in the future.
+      comp.start_date = 1.day.from_now.strftime("%F")
+      comp.end_date = 1.day.from_now.strftime("%F")
```

Comment:
Actually, with that change, it breaks our unpack_dates madness:

```
  1) DelegateReportsController logged in as THE delegate can edit report before comp is over, but cannot post report
     Failure/Error: comp.save!
     TypeError:
       no implicit conversion of Date into String
     # ./app/models/competition.rb:417:in `match'
     # ./app/models/competition.rb:417:in `unpack_dates'
     # ./spec/controllers/delegate_reports_controller_spec.rb:57:in `block (3 levels) in <top (required)>'
     # -e:1:in `<main>'

Finished in 1 minute 4.48 seconds (files took 0.83429 seconds to load)
524 examples, 1 failure

Failed examples:

rspec ./spec/controllers/delegate_reports_controller_spec.rb:53 # DelegateReportsController logged in as THE delegate can edit report before comp is over, but cannot post report
```

I'm going to leave this as is for now. Someday we'll clean up the Competitions table schema...